store/postgres/src/mapping.rs

```diff
@@ -0,0 +1,1081 @@
+pub(crate) const PRIMARY_KEY_COLUMN: &str = "id";
+
+pub(crate) const BLOCK_RANGE: &str = "block_range";
+
+#[derive(Clone, Debug)]
+pub struct Table {
+    /// The name of the GraphQL object type ('Thing')
+    pub object: s::Name,
+    /// The name of the database table for this type ('thing'), snakecased
+    /// version of `object`
+    pub name: SqlName,
+
+    pub columns: Vec<Column>,
+    /// The position of this table in all the tables for this mapping; this
+    /// is really only needed for the tests to make the names of indexes
+    /// predictable
+    position: u32,
+}
+
+impl Table {
+    fn new(
+        defn: &s::ObjectType,
+        interfaces: &mut HashMap<String, Vec<SqlName>>,
+        id_type: IdType,
+        position: u32,
+    ) -> Result<Table, StoreError> {
+        let table_name = SqlName::from(&*defn.name);
+        let columns = defn
+            .fields
+            .iter()
+            .map(|field| Column::new(field, id_type))
+            .collect::<Result<Vec<_>, _>>()?;
+        let table = Table {
+            object: defn.name.clone(),
+            name: table_name.clone(),
+            columns,
+            position,
+        };
+        for interface_name in &defn.implements_interfaces {
+            match interfaces.get_mut(interface_name) {
+                Some(tables) => tables.push(table.name.clone()),
+                None => {
+                    return Err(StoreError::Unknown(format_err!(
+                        "unknown interface {}",
+                        interface_name
+                    )))
+                }
+            }
+        }
+        Ok(table)
+    }
+    /// Find the column `name` in this table. The name must be in snake case,
+    /// i.e., use SQL conventions
+    pub fn column(&self, name: &SqlName) -> Result<&Column, StoreError> {
+        self.columns
+            .iter()
+            .find(|column| &column.name == name)
+            .ok_or_else(|| StoreError::UnknownField(name.to_string()))
+    }
+
+    /// Find the column for `field` in this table. The name must be the
+    /// GraphQL name of an entity field
+    pub fn column_for_field(&self, field: &str) -> Result<&Column, StoreError> {
+        self.columns
+            .iter()
+            .find(|column| &column.field == field)
+            .ok_or_else(|| StoreError::UnknownField(field.to_string()))
+    }
+
+    pub fn primary_key(&self) -> Reference {
+        Reference {
+            table: self.name.clone(),
+            column: PRIMARY_KEY_COLUMN.into(),
+        }
+    }
+
+    pub fn reference(&self, name: &SqlName) -> Result<Reference, StoreError> {
+        let attr = self.column(name)?;
+        Ok(Reference {
+            table: self.name.clone(),
+            column: attr.name.clone(),
+        })
+    }
+
+    fn as_ddl(&self, out: &mut String, mapping: &Mapping) -> fmt::Result {
+        write!(out, "create table {}.{} (\n", mapping.schema, self.name)?;
```

Comment:
Would be good to quote the identifiers.

Reply:
I decided to check the names we take in in `Mapping` for SQL compatibility rather than quoting things everywhere. In practice, it shouldn't matter for users as the naming rules match GraphQL well, but it makes it clear in the code that not quoting those names is ok. Commit 9456efdf

Reply:
Great!